Optimize topology::compare_endpoints

bhalevy opened the pull request:
The code for compare_endpoints originates at the dawn of time (bc034aeaecb68681e5c7c26f4cbda00f14aa27c0)
and is called on the fast path from storage_proxy via `sort_by_proximity`.

This series considerably reduces the function's footprint by:
1. carefully coding the many comparisons in the function so to reduce the number of conditional banches (apparently the compiler isn't doing a good enough job at optimizing it in this case)
2. avoid sstring copy in topology::get_{datacenter,rack}

xemul commented:
Regarding the 2nd patch -- how can we sure the obtained dc/rack string is not used through preemption point? Currently it survives because its just a copy of a string, after the patch it can be come dangling reference

xemul commented on locator/topology.cc:
```diff
@@ -167,15 +167,18 @@ void topology::sort_by_proximity(inet_address address, inet_address_vector_repli
 }
 
 int topology::compare_endpoints(const inet_address& address, const inet_address& a1, const inet_address& a2) const {
```
I believe this particular comparer deserves its own personal unit test

bhalevy replied:
Maybe, though it's used on the fast path so any cluster test exercises it, in particular multi-dc ones.

avikivity replied:
I think mistakes here won't show up in tests, as they end up sending requests to a node that is too far away rather than a node that doesn't hold data.

bhalevy replied:
I'm adding a unit test as we speak :)

bhalevy commented:
> Regarding the 2nd patch -- how can we sure the obtained dc/rack string is not used through preemption point? Currently it survives because its just a copy of a string, after the patch it can be come dangling reference

It's the callers problem, like many getters that return a reference to a member the caller needs to copy the value if they need to extend its lifetime across continuations.

xemul commented:
> > Regarding the 2nd patch -- how can we sure the obtained dc/rack string is not used through preemption point? Currently it survives because its just a copy of a string, after the patch it can be come dangling reference
> 
> It's the callers problem, like many getters that return a reference to a member the caller needs to copy the value if they need to extend its lifetime across continuations.

Of course it's callers problem, my question is -- what the chances that there are no such places in code and is there anything that can be done to spot them?

bhalevy commented:
> > > Regarding the 2nd patch -- how can we sure the obtained dc/rack string is not used through preemption point? Currently it survives because its just a copy of a string, after the patch it can be come dangling reference
> > 
> > 
> > It's the callers problem, like many getters that return a reference to a member the caller needs to copy the value if they need to extend its lifetime across continuations.
> 
> Of course it's callers problem, my question is -- what the chances that there are no such places in code and is there anything that can be done to spot them?

ASAN to the rescue

avikivity commented:
> The code for compare_endpoints originates at the dawn of time ([bc034ae](https://github.com/scylladb/scylladb/commit/bc034aeaecb68681e5c7c26f4cbda00f14aa27c0)) and is called on the fast path from storage_proxy via `sort_by_proximity`.
> 
> This series considerably reduces the function's footprint by:

How was this measured?

> 
> 1. carefully coding the many comparisons in the function so to reduce the number of conditional banches (apparently the compiler isn't doing a good enough job at optimizing it in this case)
> 2. avoid sstring copy in topology::get_{datacenter,rack}

avikivity commented:
> > The code for compare_endpoints originates at the dawn of time ([bc034ae](https://github.com/scylladb/scylladb/commit/bc034aeaecb68681e5c7c26f4cbda00f14aa27c0)) and is called on the fast path from storage_proxy via `sort_by_proximity`.
> > This series considerably reduces the function's footprint by:
> 
> How was this measured?

Ah, patches add detail.

avikivity commented on locator/topology.cc:
```diff
@@ -200,18 +202,25 @@ int topology::compare_endpoints(const inet_address& address, const inet_address&
-    //
-    // We don't differentiate between Nodes if all Nodes belong to different
-    // Data Centers, thus make them equal.
-    //
```
Idea (needs node_ptr first): give each node a metric `(dc_seq << 48) | (rack_seq << 32) | node_seq`. The actual sequence numbers don't matter.

Then, this function is abs(n1.metric - n.metric) <=> (n2.metric - n.metric).

bhalevy replied:
It won't be exactly backward compatible for nodes in different racks in same dc, but I think it's work just as well

bhalevy replied:
or even nodes in the same rack, for both cases we return 0 today, but 0 here stands for "don't care" as as I understood from the comments.

bhalevy replied:
I think this would be a nice follow up over https://github.com/scylladb/scylladb/pull/11987
where we can keep and get the metric in the node object.
In the current codebase we can extend topology or endpoint_dc_rack to compute and keep the metric, but I'd rather spend the effort in the nodes model.

avikivity replied:
Since the metric is local, we could assign the same value for remote dc/demote racks.

`metric = !is_same_node | (!is_same_rack << 1) | (!is_same_dc << 1)`

7 = very far, 3 = local dc, 1 = same rack, 0 = looking in mirror

avikivity replied:
and, of course, it could be precalculated.

avikivity commented on locator/topology.hh:
```diff
@@ -71,10 +71,10 @@ public:
+    const sstring& get_rack(inet_address ep) const;
+    const sstring& get_datacenter() const;
+    const sstring& get_datacenter(inet_address ep) const;
 
```
Even better would be to convert them to pointers to rack/dc objects in token_metadata (then compare pointers), but of course a much bigger change.

bhalevy replied:
And should be based on https://github.com/scylladb/scylladb/pull/11987 where node objects would be stable in the topology instance (and stable forever on shard 0)

avikivity commented:
Looks good, but agree a regression test is needed.

I'm not very worried about returning references - usually they'll be captured by value and so a copy is created anyway. And the caller will hold the metadata reference.

bhalevy commented:
In [cc55415](https://github.com/scylladb/scylladb/commit/cc55415d31247364906fb77149e722bba5c092bc): added test_topology_compare_endpoints unit test

avikivity commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -631,3 +631,80 @@ SEASTAR_TEST_CASE(test_invalid_dcs) {
+    const auto& address = nodes[tests::random::get_int<size_t>(0, NODES-1)];
+    const auto& a1 = nodes[tests::random::get_int<size_t>(0, NODES-1)];
+    const auto& a2 = nodes[tests::random::get_int<size_t>(0, NODES-1)];
+
```
If the test uses randomness, then it should repeat. There's a 1:1000 chance it's testing a, a, a.

bhalevy replied:
Actually that chance is 1:1, see below statement.
I think repeating can be done externally, and it is also achieved naturally when the test is run in CI, next promotion, and daily testing jobs.

avikivity replied:
Ok. I don't think repeats are needed if you test all cases.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4236/

bhalevy commented:
In v3 ([6afa608](https://github.com/scylladb/scylladb/commit/6afa608e0550bf4ec7b57f5f903f528d732e2da1)):
- topology: add node location metric
  - precalculated, based on https://github.com/scylladb/scylladb/pull/12761#discussion_r1098529840
- topology: optimize compare_endpoints
  - compiler now inlines compare_endpoints into sort_by_proximity
- dropped "topology: let get_{dc,rack} return result by ref"
  - will be sent separately

bhalevy commented on locator/topology.hh:
```diff
@@ -101,7 +102,7 @@ private:
      * Comparator.compare would
      */
-    int compare_endpoints(const inet_address& address, const inet_address& a1, const inet_address& a2) const;
+    std::strong_ordering compare_endpoints(const inet_address& address, const inet_address& a1, const inet_address& a2) const;
```
fair point.  weak ordering indeed better describes the requirements for this function.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4248/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4250/

bhalevy commented:
In v5 ([dcca59d](https://github.com/scylladb/scylladb/commit/dcca59d17252ad30473d0ceff161a24dfaffd34f)):
- addressed exception seen in CI
- rename _node_location_map member to _node_metrics
- compare_endpoints now implicitly creates a default entry in _node_metric, equivalent to using the default endpoint_dc_rack previously.
- added testing for a bogus address in unit test

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4255/

bhalevy commented:
@scylladb/scylla-maint ping

avikivity commented on locator/topology.cc:
```diff
@@ -73,8 +79,21 @@ void topology::update_endpoint(const inet_address& ep, endpoint_dc_rack dr)
+
+    location_metric m = 0;
+    auto [dc_it, dc_emplaced] = _dc_radix.try_emplace(dr.dc, _dc_radix.size());
+    location_metric dc_metric = dc_it->second;
```
I think it's better to use just 0/1 (and allocate just one bit for each component). This way, two different datacenters or two different nodes in the same rack are the same distance away and we don't favor one dc over another based on some irrelevant criteria. The code becomes simpler too, because you don't need to assign different nonzero values.

avikivity replied:
Actually, that only works if sort_by_proximity is only called for this node (why wouldn't it?)

In all but one case I could verify that this is the case, probably the last one is too.

bhalevy replied:
If we want to preserve that DCs different from the reference node DC are indistinguishable and similarly, rack different than the reference rack in the same DC, we need to compute the metric on each call, which I wanted to avoid, in order to reduce the number of logical branches.

Since the language does guarantee that bool to int conversion will produce 0 or 1 (See https://en.cppreference.com/w/cpp/language/implicit_conversion), we can convert the boolean comparison value of dc/rack to single bit with branching.

avikivity commented on to_string.hh:
```diff
@@ -163,4 +163,28 @@ static inline std::ostream& operator<<(std::ostream& os, const std::strong_order
+        os << "eq";
+    }
+    return os;
+}
```
Please deinline.

avikivity replied:
(it's not really up to us for format standard stuff, but don't see an alternative)

michoecho replied:
The alternative is to wrap the standard type into our own type and use that instead, (and provide an implicit conversion back to the standard type, if that's needed).

avikivity commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -634,43 +635,43 @@ SEASTAR_TEST_CASE(test_invalid_dcs) {
 
 void topology::test_compare_endpoints(const inet_address& address, const inet_address& a1, const inet_address& a2) const {
-    std::optional<int> expected;
+    auto expected = std::partial_ordering::unordered;
```
How can we have unordered here? I understand equivalent/less/greater, but what is unordered here?

michoecho replied:
It expresses that the comparison function is permitted to choose the ordering arbitrarily (meaning that the test doesn't enforce the choice) in some cases (here: when the candidate nodes are both in the same DC, but different racks than the source node).

Note that the comparison function returns `weak_ordering`, not `partial_ordering`, so it's never `unordered`. The test just uses `partial_ordering::unordered` as a "semantically flavoured" version of `optional<weak_ordering>`, where `expected == nullopt` means that we don't care about the result.

bhalevy replied:
exactly

bhalevy replied:
I can keep it as optional in the test if it makes things any clearer.

bhalevy commented:
In v6 ([bb36237](https://github.com/scylladb/scylladb/commit/bb36237cf4b56385614eb51bfe1038b6816ff7a3)):
- test: network_topology: add test_topology_compare_endpoints
  - added some comments about the expected proximity order
- Added:
  - move to_string.hh to utils/
  - utils: to_sstring: deinline std::strong_ordering print operator
  - and added printers for weak/partial ordering in utils/to_string.cc
- Dropped "topology: add node location metric"
- Instead, calculating distance metric on the fly in topology::compare_endpoints
  - this still considerably shrinks the function footprint, though not below the inlining threshold
  - and gets rid of many logic branches
  - using object pointers for dc/rack instead of strings in the future should make them even cheaper to compare

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4373/

bhalevy commented:
@avikivity @michoecho ping, please re-review.

bhalevy commented:
@avikivity ping, please re-review and merge if fine
